### DISCUSSION TITLE
Make Tasmota Y2038 secure?

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#19590] [Ideas]
Until 2038, it's only 15 years now. 

Unix timestamp will "wind around" to -1, which is 1901 on most devices.

While many might think "ah that's a loong time to go, every electrical plug will have been replaced by then" let me reply "OK fine but if we wait another 10 years, how would we then address the problem?

For example, is the problem "solved" already as Tasmota is using a nonnegative integer on all such timestamps? If so, has someone tested it?

We are dealing with timestamps in many places here, so are they ALL nonnegative ints?

Or does it make more sense to switch to 64 bit? Or can we come up with 40 bit?

I'd just like the developers to maybe drop some lines on that, that might suffice if someone researches on that topic in the future.

**Comment by s-hadinger** [comment]
https://github.com/arendst/Tasmota/blob/0f6b59f67e45c4b79b6f8475c36b098d8fd2e81d/tasmota/tasmota_support/support_rtc.ino#L39C1-L54C7

```
struct RTC {
  uint32_t utc_time = 0;
  uint32_t local_time = 0;
  uint32_t daylight_saving_time = 0;
  uint32_t standard_time = 0;
  uint32_t midnight = 0;
  uint32_t restart_time = 0;
  uint32_t nanos = 0;
  uint32_t millis = 0;
[...]
} Rtc;
```

So we're ok until year 2106.

**Reply by sfromis** [reply]
Not so fast.....

Epoch in seconds for 2040-01-01 is 2208985200, and Berry processes epoch in seconds, signed integer.
```
print(2208985200, tasmota.time_dump(2208985200))
-2085982096 {'min': 31, 'weekday': 3, 'sec': 44, 'month': 11, 'year': 1903, 'day': 25, 'epoch': -2085982096, 'hour': 16}
```

**Comment by hagenbuch** [comment]
Great, thanks - that was what I was aiming at. I think I'd like to test it in some spare time, but as the clock always sets itself automatically, how could we test it? Can we set the clock manually? Maybe output routines don't know about the unsigned?

**Reply by sfromis** [reply]
The `Time` command allows you to set the clock manually, epoch value in seconds.

**Reply by sfromis** [reply]
Did not work great, but that may only be parsing of the Time command....
```
13:39:44.504 CMD: time 2208985200
04:14:07.001 RSL: TIME = {"Time":"2038-01-19T04:14:07+01:00","Epoch":2147483647}
```

Setting to near the rollover, and waiting does seem to work:
```
RSL: TIME = {"Time":"2038-01-19T04:14:51+01:00","Epoch":2147483691}
```

**Reply by s-hadinger** [reply]
Berry does display as 32 bits signed indeed. But none of the time calculation is done in Berry, and epoch is always converted back to unsigned 32 bits. So as long as you are ok with the negative value displayed, and you don't compare in Berry you are fine.

We could also consider moving to int64 in Berry for epoch

**Reply by sfromis** [reply]
My concern with Berry was not about seeing a negative number printed, but about seeing year 1903 (and also all the other split values bad) instead of the expected 2040 to match the 32-bit epoch value.

**Reply by s-hadinger** [reply]
Hmmm. This is worse that I thought.

`time_t` is currently `long int` which means 32 bits signed. So our `libc` is not 2038 safe.

**Comment by hagenbuch** [comment]
I have to thank you all! 

It's fantastic how you responded and I feel a bit sorry for creating the kerfuffle but it has to be solved at a certain time anyway..

**Reply by s-hadinger** [reply]
Yes indeed. But it's still low priority
